### PR TITLE
Fix some bugs, account for custom case

### PR DESCRIPTION
**Pull request by kwalcock**

### CODE REVIEW

File: main/src/main/scala/org/clulab/dynet/ConstEmbeddingsGlove.scala [hunk 1]
```diff
@@ -42,11 +42,11 @@ object ConstEmbeddingsGlove {
     mkConstLookupParams(sentence.words.toSet)
 
   def mkConstLookupParams(doc: Document, lowerCase: Boolean = false): ConstEmbeddingParameters = {
```
Comment (kwalcock) [inline main/src/main/scala/org/clulab/dynet/ConstEmbeddingsGlove.scala:44]: This code didn't seem to be taking the value of `lowerCase` into account.

File: main/src/main/scala/org/clulab/processors/clu/CluProcessor.scala [hunk 2]
```diff
@@ -235,47 +234,45 @@ class CluProcessor protected (
...
-      for(sent <- doc.sentences) {
-        // the case restoration model expects lower-case words as input                    
-        val loweredWords = sent.words.map(_.toLowerCase())
+      for (sent <- doc.sentences) {
+        // The case restoration model expects lower-case words as input.
+        val originalWords = sent.words
+        val loweredWords = originalWords.map(_.toLowerCase)
         val preLabels = mtlCase.predict(0, AnnotatedSentence(loweredWords), None, getEmbeddings(doc))
         val labels = casePostProcessing(loweredWords, preLabels)
-        assert(labels.size == loweredWords.size)
```
Comment (kwalcock) [inline main/src/main/scala/org/clulab/processors/clu/CluProcessor.scala:243]: There should be a stronger guarantee in `casePostProcessing` that this is the case.

File: main/src/main/scala/org/clulab/processors/clu/CluProcessor.scala [hunk 3]
```diff
@@ -235,47 +234,45 @@ class CluProcessor protected (
...
-        val restoredWords = loweredWords.zip(labels).map(x => restoreCaseWord(x._1, x._2))
-        for(i <- sent.indices) {
```
Comment (kwalcock) [inline main/src/main/scala/org/clulab/processors/clu/CluProcessor.scala:245]: There seems to be a library function for this.

File: main/src/main/scala/org/clulab/processors/clu/CluProcessor.scala [hunk 4]
```diff
@@ -235,47 +234,45 @@ class CluProcessor protected (
...
+        restoredWords.copyToArray(originalWords)
       }
     }                    
   }
 
   private def casePostProcessing(loweredWords: IndexedSeq[String], preLabels: IndexedSeq[String]): IndexedSeq[String] = {
-    val labels = new ArrayBuffer[String]
-    for(i <- loweredWords.indices) {
-      val w = loweredWords(i)
-      for(pl <- CASE_PATTERNS) {
```
Comment (kwalcock) [inline main/src/main/scala/org/clulab/processors/clu/CluProcessor.scala:256]: If there are multiple `CASE_PATTERNS` that match, multiple labels will be added and the sizes of the sequences will not be equal, I think.

File: main/src/main/scala/org/clulab/processors/clu/CluProcessor.scala [hunk 5]
```diff
@@ -235,47 +234,45 @@ class CluProcessor protected (
...
+
+  private def restoreCaseWord(loweredWord: String, label: String, originalWord: String): String = {
+    val alphaCount = originalWord.count(_.isLetter)
+    val upperCount = originalWord.count(_.isUpper)
+    val isStandard =
+        upperCount == 0 || // lower
+        upperCount == alphaCount || // all upper
+        upperCount == 1 && originalWord.head.isUpper // upper initial
+    // We handle three possible labels: L (lower), UI (upper initial), UA (all upper).
+    if (!isStandard)
```
Comment (kwalcock) [inline main/src/main/scala/org/clulab/processors/clu/CluProcessor.scala:268]: This tries to account for custom case usage.

**Review by kwalcock — COMMENTED** [review]
I added some explanation.

**Comment by kwalcock** [comment]
There's a problem with the Jenkins-Artifactory communication that I'm working on which is causing the failed tests.

**Comment by kwalcock** [comment]
| Original text | Previous restoration | Proposed restoration |
| --- | --- | --- |
| Saint-Louis | Saint-louis | Saint-Louis |
| POs | POS | POs |
| LBA-East | Lba-east | LBA-East |
| BdD- | BDD- | BdD- |

**Comment by kwalcock** [comment]
[Bulletin_02_06_2020.en.txt.preserved.txt](https://github.com/clulab/processors/files/8543288/Bulletin_02_06_2020.en.txt.preserved.txt)

[previous-Bulletin_02_06_2020.en.txt.restored.txt](https://github.com/clulab/processors/files/8543286/previous-Bulletin_02_06_2020.en.txt.restored.txt)

[proposed-Bulletin_02_06_2020.en.txt.restored.txt](https://github.com/clulab/processors/files/8543290/proposed-Bulletin_02_06_2020.en.txt.restored.txt)

**Comment by MihaiSurdeanu** [comment]
LGTM @kwalcock !
Thanks for handling it. 
Ok to merge.

**Pull request merged**